pep-0572.rst

```diff
@@ -62,140 +57,151 @@ that value::
     # Share a subexpression between a comprehension filter clause and its output
     filtered_data = [y for x in data if (y := f(x)) is not None]
 
+Scope of the target
+-------------------
 
-Differences from regular assignment statements
-----------------------------------------------
+An assignment expression does not introduce a new scope.  In most
+cases the scope in which the target will be bound is self-explanatory:
+it is the current scope.  If this scope contains a ``nonlocal`` or
+``global`` declaration for the target, the assignment expression
+honors that.
+
+There is one special case: an assignment expression occurring in a
+list, set or dict comprehension or in a generator expression (below
+collectively referred to as "comprehensions") binds the target in the
+containing scope, honoring a ``nonlocal`` or ``global`` declaration
+for the target in that scope, if one exists.  For the purpose of this
+rule the containing scope of a nested comprehension is the scope that
+contains the outermost comprehension.
```

Comment:
That wasn't my intent (I intended "in the parent block,  period") - but this variation makes more sense to me - thanks!  So long as expressions can't contain statements, it makes things "just work" more often, at the cost of making it impossible for inner comprehensions to rebind outer comprehensions'  `for` target names.  That's no real loss.

If, however, Python grows to support "expressions" that can contain arbitrary statements, then it also makes it impossible for comprehensions to rebind other kinds of local names too in outer comprehensions.

As is, this example is annoying ;-)

    def f():
        global j
        [[(j := j) for i in range(5)] for j in range(5)]

In `j := j`, is the left `j` global and the right `j` local to the outer listcomp?  We already have (or should have) a rule that makes it an error for a name to be both a `for` and an `:=` target at the _same_ level, but the example dodges that rule.  In the "always in the parent block" version, both instances of `j` refer to the outer listcomp's `j`.

"Left is global, right is local to containing listcomp" is fine by me in practice, as would be just about any other resolution..  But it merits some thought.

Reply:
Oh, I forgot about the bit where we agreed that you can't rebind the loop control variable of a comprehension using `:=`. Let's extend that so that you can't rebind the loop control variable of any containing comprehension. That would make this example invalid, which seems better than allowing the example to dodge the rule.

Reply:
That works for me.  There's a formal property I'm keen to preserve:  that at any given comprehension level, all instances of a `:=` target resolve to the same scope.  So if any other seeming counterexamples come up, I expect I'd rather say "you can't do that" than pick one more-than-less arbitrarily.